File: src/main/java/OneTimePad.java
import java.security.SecureRandom;

class OneTimePad {
  private static SecureRandom secureRandom = new SecureRandom();
  private static String message = "Meet me at secret location at noon on Wednesday.";
  private static int ALPHABET_SIZE = 94; //ASCII 32 - 126
  private static int PAD_SIZE = 100;

  public static void main(String[] args) {
    final char[] plaintext = message.toCharArray();

    // Generate secure random key (never reuse this)
    final int[] pad = new int[PAD_SIZE];
    for (int i = 0; i < pad.length; i++) {
      pad[i] = secureRandom.nextInt(ALPHABET_SIZE);
    }

    // Combine the key with the plaintext
    char[] ciphertext = new char[PAD_SIZE];
    for (int i = 0; i < PAD_SIZE; i++) {
      if (i < plaintext.length)
        ciphertext[i] = getShiftChar(plaintext[i], pad[i]);
      else
        ciphertext[i] = getShiftChar(' ', pad[i]);
    }

    System.out.println("Ciphertext: " + new String(ciphertext));

    // Separate the key from the ciphertext
    char[] decrypted = new char[PAD_SIZE];
    for (int i = 0; i < PAD_SIZE; i++) {
      decrypted[i] = getShiftChar(ciphertext[i], -pad[i]);
    }

    System.out.println("Decrypted text: " + new String(decrypted));
  }

  static char getShiftChar(char c, int shift) {
    int x = c + shift - 32;
    int z = Math.floorMod(x, ALPHABET_SIZE);
    char i = (char) (z + 32);
    return i;
  }
}

File: src/main/kotlin/ciphers/OneTimePad.kt
package ciphers

import java.security.SecureRandom

/**
 * An OTP mixes the ciphertext with a long random key. Never use this key twice!
 */

private val secureRandom = SecureRandom()
private val ALPHABET_SIZE = 94 //ASCII 32 - 126
private val PAD_SIZE = 100

fun main(args: Array<String>) {
  val message = "Meet me at secret location at noon on Wednesday."
  println("Plaintext:    " + message)

  // Generate secure random key (never reuse this)
  val pad = ByteArray(PAD_SIZE)
  secureRandom.nextBytes(pad)
  println("One time pad: " + pad.map { getShiftChar(0.toChar(), it) } .joinToString(""))

  val ciphertext = encrypt(message, pad)
  println("Ciphertext:   " + ciphertext)

  val decrypted = decrypt(ciphertext, pad)
  println("Decrypted:    " + decrypted)
  println(Byte.MAX_VALUE.toChar())
}

private fun encrypt(plaintext: String, pad: ByteArray): String {
  // Combine the key with the plaintext
  return plaintext.mapIndexed { i, c -> getShiftChar(c, pad[i]) }.joinToString("")
}

private fun decrypt(ciphertext: String, pad: ByteArray): String {
  // Separate the key from the ciphertext
  return ciphertext.mapIndexed { i, c -> getShiftChar(c, (-pad[i]).toByte()) }.joinToString("")
}

private fun getShiftChar(c: Char, shift: Byte): Char {
  val x = c.toInt() - 32 + shift
  val z = Math.floorMod(x, ALPHABET_SIZE)
  val i = (z + 32).toChar()
  return i
}